File: 02x10/c-012/main.rb
def factorsLengthOver num, maxnum
    totalFactor = 0
    limit = num
    i = 1
    while i < limit do
        if num % i == 0 then
            if num / i == i then
                totalFactor += 1
            else
                totalFactor += 2
            end
            limit = num / i
            if totalFactor > maxnum then
                return true
            end
        end
        i += 1
    end
    false
end

def triangleFactorOver maxFactor
    triangle = 0
    i = 1
    loop do
        triangle += i
        break if factorsLengthOver(triangle, maxFactor)
        i += 1
    end
    triangle
end

puts triangleFactorOver(500)

File: 02x10/c-012/main.go
package main

import "fmt"

func factorsLengthOver(num int, max int) bool {
	totalFactor := 0
	limit := num
	for i := 1; i < limit; i++ {
		if num%i == 0 {
			if num/i == i {
				totalFactor += 1
			} else {
				totalFactor += 2
			}
			limit = num / i
			if totalFactor > max {
				return true
			}
		}
	}
	return false
}

func triangleFactorOver(maxFactor int) int {
	triangle := 0
	for i := 1; ; i++ {
		triangle += i
		if factorsLengthOver(triangle, maxFactor) {
			break
		}
	}
	return triangle
}

func main() {
	fmt.Println(triangleFactorOver(500))
}
